### DISCUSSION TITLE
Bug - setValue + fieldArray not working as expected (only with Controller)

### DISCUSSION THREAD

**Discussion by danielbattat** [react-hook-form/react-hook-form#3532] [Q&A] [answered]
It seems that when using Controller (I use Antd), the watched value for a fieldArray does not update correctly when using setValue.

The first iteration of watch returns some undefined values which can cause issues for obvious reasons. It seems that when using register everything works fine.

Controller: https://codesandbox.io/s/usefieldarray-with-setvalue-forked-ijykf?file=/src/App.js
See console:
![image](https://user-images.githubusercontent.com/30219620/100401720-f88f4800-3062-11eb-9132-de9b6cdea81a.png)

Register (works fine):
https://codesandbox.io/s/usefieldarray-with-setvalue-forked-u1i5w?file=/src/App.js
![image](https://user-images.githubusercontent.com/30219620/100401784-283e5000-3063-11eb-9630-0ce79fa919ed.png)

Is this a bug or am I missing something. Alternatively, is it possible to use Antd without a Controller? When I try it seems to work but I get the warning "Field is missing `name` attribute".

Thank you!

**Comment by bluebill1049** [comment, marked as answer]
you may want to use the latest version: https://codesandbox.io/s/usefieldarray-with-setvalue-forked-g3uuf?file=/src/App.js

**Reply by danielbattat** [reply]
Thank you so much for the quick reply. Interestingly, I *am* using 6.11.5 on my computer and I experienced this issue in my code. I will try to understand what is going wrong, tomorrow.

**Reply by bluebill1049** [reply]
no worries, if you can reproduce the issue in a CSB, happy to investigate for a solution.

**Reply by danielbattat** [reply]
Hi @bluebill1049 , I was able to reproduce in CSB: 
https://codesandbox.io/s/usefieldarray-with-setvalue-forked-fjq42

The issue happens here if you add a record using append, and then use the ModalEditor to update the value of "c". The first iteration of the useEffect only has {c} in the object, causing issues when the useEffect expects the whole object.

For some reason calling "update" manually does not cause this issue.

Am I doing something wrong?

**Reply by bluebill1049** [reply]
hey @danielbattat thanks for the CSB, it certainly helps. What's the correct steps to reproduce the issue?

**Reply by bluebill1049** [reply]
i did some clean up: https://codesandbox.io/s/usefieldarray-with-setvalue-forked-4t8gp?file=/src/App.js

**Reply by danielbattat** [reply]
Thank you.
Step 1. Click "Add"
Step 2. Click "Open Editor", type a value for "c" and click "save c".
Step 3. Notice that console.log of nest is missing "a" and "b":

![image](https://user-images.githubusercontent.com/30219620/100528564-6bf0a100-31ac-11eb-90f2-091521430b30.png)

**Comment by bluebill1049** [comment]
I think i finally find a solution for it: https://codesandbox.io/s/usefieldarray-with-setvalue-forked-4t8gp?file=/src/src/useField.ts

**Reply by danielbattat** [reply]
Hi @bluebill1049 
Thank you for your effort. The issue is still happening in your csb; 
![image](https://user-images.githubusercontent.com/30219620/100538472-b586cd80-3238-11eb-895a-ab5c6d206b67.png)

**Reply by bluebill1049** [reply]
check out the latest patch: https://codesandbox.io/s/usefieldarray-with-setvalue-forked-4t8gp?file=/src/App.js

<img width="584" alt="Screen Shot 2020-11-29 at 8 59 20 pm" src="https://user-images.githubusercontent.com/10513364/100538757-c27dde00-3285-11eb-8422-f96f0ecad6b9.png">

unless i am missing the steps.

**Reply by danielbattat** [reply]
This works!!! Can you let me know what you changed so I can fix it on my end? Thank you!!

**Reply by bluebill1049** [reply]
you don't have to change anything, just pull the latest patch.

**Reply by danielbattat** [reply]
Thank you, so grateful for your library and quick help!

**Reply by bluebill1049** [reply]
pleasure mate.

**Reply by bluebill1049** [reply]
There is an issue with the fix above, i am patching it, so make sure you update it once its out.

**Reply by bluebill1049** [reply]
```
setValue(
        "nest",
        [...nest.map((n) => ({ ...n, ...payload }))]
      );
```

This should fix the issue without any of the new patches for you.

**Reply by danielbattat** [reply]
Hi @bluebill1049 
Thank you for following up. I upgraded to 6.12.2 and did not make any change to my code, and it still works. This is what I have in my code:

```
setValue(
        "nest",
        nest.map((n) => ({ ...n, ...payload }))
      );
```

I saw that you reverted the fix that you did, so I am not sure why its working now. Is this what you expected, or did you want me to add the [... nest]?
